File: src/pers/gargantua/mine_sweep/Start.java
package pers.gargantua.mine_sweep;

import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;
import java.awt.event.ActionEvent;

import javax.swing.JButton;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JTextField;

public final class Start extends JPanel{

	private static Start start = null;
	
	private final JTextField row;
	private final JTextField column;
	private final JTextField mines;
	
	
	public static Start getStart() {
		if(start == null)
			start = new Start();
		return start;
	}
	
	private Start() {
		
		GridBagLayout gridBagLayout = new GridBagLayout();
		GridBagConstraints first = new GridBagConstraints();
		GridBagConstraints last = new GridBagConstraints();
		
		setLayout(gridBagLayout);
		
		first.fill = GridBagConstraints.BOTH;
		last.fill = GridBagConstraints.BOTH;
		last.gridwidth = GridBagConstraints.REMAINDER;
		
		JLabel r = new MyLabel("Row :", gridBagLayout, first);
		JLabel c = new MyLabel("Column :", gridBagLayout, first);
		JLabel m = new MyLabel("Mines :", gridBagLayout, first);
		
		row = new MyText("", gridBagLayout, last);
		column = new MyText("", gridBagLayout, last);
		mines = new MyText("", gridBagLayout, last);
		
		JButton easy = new MyButton("EASY", gridBagLayout, first);
		JButton ord = new MyButton("MIDDLE", gridBagLayout, first);
		JButton hard = new MyButton("HARD", gridBagLayout, last);
		JButton startButton = new MyButton("Game Start!", gridBagLayout, last);
		
		easy.addActionListener((ActionEvent e) -> setting(Config.EASY));
		ord.addActionListener((ActionEvent e) -> setting(Config.ORD));
		hard.addActionListener((ActionEvent e) -> setting(Config.HEAD));
		startButton.addActionListener((ActionEvent e) -> {
			int row;
			int column;
			int mine;
			try {
				row = Integer.parseInt(this.row.getText());
				column = Integer.parseInt(this.column.getText());
				mine = Integer.parseInt(mines.getText());
			}catch (NumberFormatException e1) {
				new MyDialog("Invalid Value", "ERROR", false);
				return;
			}
			if (!(this.row.getText().equals("") || this.column.getText().equals("") || mines.getText().equals(""))) {
				if(row < 1 || column < 1){
					new MyDialog("Too few rows or columns", "ERROR", false);
				} else if(mine < 1){
					new MyDialog("Too few mines", "ERROR", false);
				} else if (row * column < mine) {
					new MyDialog("Too many mines", "ERROR", false);
				} else {
					MainFrame.getMainFrame().cardLayout.show(MainFrame.getMainFrame().mainPanel, "gaming");
					Maps.getMaps().config = new Config(row, column, mine);
					Gaming.getGaming().start();
					StaBar.getStaBar().mines = 0;
					StaBar.getStaBar().times = 0;
					StaBar.getStaBar().run();
				}
			}
		});
		
		add(r);
		add(row);
		add(c);
		add(column);
		add(m);
		add(mines);
		add(easy);
		add(ord);
		add(hard);
		add(startButton);
		
	}

	public void setting(Config config) {
		row.setText(Integer.toString(config.r));
		column.setText(Integer.toString(config.c));
		mines.setText(Integer.toString(config.m));
	}
}

File: src/pers/gargantua/mine_sweep/Start.kt
package pers.gargantua.mine_sweep

import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import javax.swing.JPanel

/**
 * @author Gargantua丶
 **/
object Start : JPanel() {

    private val gridBagLayout = GridBagLayout()
    private val first = GridBagConstraints().apply {
        fill = GridBagConstraints.BOTH
    }
    private val last = GridBagConstraints().apply {
        fill = GridBagConstraints.BOTH
        gridwidth = GridBagConstraints.REMAINDER
    }

    private val r = MyLabel("Row :", gridBagLayout, first)
    private val c = MyLabel("Column :", gridBagLayout, first)
    private val m = MyLabel("Mines :", gridBagLayout, first)

    private val row = MyText("", gridBagLayout, last)
    private val column = MyText("", gridBagLayout, last)
    private val mines = MyText("", gridBagLayout, last)

    private val easy = MyButton("EASY", gridBagLayout, first).apply {
        addActionListener {
            setting(PresetConfig.EASY)
        }
    }
    private val ord = MyButton("ORD", gridBagLayout, first).apply {
        addActionListener {
            setting(PresetConfig.ORD)
        }
    }
    private val hard = MyButton("HEAD", gridBagLayout, last).apply {
        addActionListener {
            setting(PresetConfig.HEAD)
        }
    }
    private val start = MyButton("Game Start!", gridBagLayout, last).apply {
        addActionListener {
            var res = true
            var r = 0
            var c = 0
            var m = 0
            try {
                r = row.text.toInt()
                c = column.text.toInt()
                m = mines.text.toInt()
            } catch (e: NumberFormatException) {
                res = false
                MyDialog("ERROR", "Invalid Value", sta = false)
            }
            if(res) {
                if(r < 1 || c < 1){
                    MyDialog("ERROR", "Too few rows or columns", sta = false)
                } else if(m < 1){
                    MyDialog("ERROR", "Too few mines", sta = false)
                } else if (r * c < m) {
                    MyDialog( "ERROR", "Too many mines", sta = false)
                } else {
                    MainFrame.cardLayout.show(MainFrame.mainPanel, "gaming")
                    Maps.config = Config(r, c, m)
                    Gaming.start()
                    StaBar.mines = 0
                    StaBar.times = 0
                    StaBar.run()
                }
            }
        }
    }

    init {
        layout = gridBagLayout

        add(r)
        add(row)
        add(c)
        add(column)
        add(m)
        add(mines)
        add(easy)
        add(ord)
        add(hard)
        add(start)
    }

    private fun setting(config: Config) {
        row.text = config.r.toString()
        column.text = config.c.toString()
        mines.text = config.m.toString()
    }

}
